applications/FluidDynamicsApplication/custom_processes/shock_capturing_entropy_viscosity_process.cpp

```diff
@@ -0,0 +1,390 @@
+        r_node.SetValue(ARTIFICIAL_BULK_VISCOSITY, 0.0);
+    });
+}
+
+
+void ShockCapturingEntropyViscosityProcess::ComputeArtificialMagnitudes()
+{
+    KRATOS_TRY
+
+    const double delta_time = mrModelPart.GetProcessInfo().GetValue(DELTA_TIME);
+
+    if(mrModelPart.ElementsBegin() == mrModelPart.ElementsEnd()) return; // Empty modelpart
+
+    const double heat_capacity_ratio = mrModelPart.ElementsBegin()->GetProperties().GetValue(HEAT_CAPACITY_RATIO);
+
+    block_for_each(mrModelPart.Elements(), [&](Element& r_element)
+    {
+        const auto inf_norm = ComputeElementalInfNormData(r_element, delta_time, heat_capacity_ratio);
+        const double h2 = ComputeHSquared(r_element);
+
+        r_element.SetValue(SHOCK_SENSOR, inf_norm.EntropyResidual);
+
+        const double mu_e = mEntropyConstant * h2 * inf_norm.Density * inf_norm.EntropyResidual;
+        const double mu_max = mEnergyConstant * std::sqrt(h2) * inf_norm.Density * inf_norm.TotalVelocity;
+        const double mu_h = std::min(mu_e, mu_max);
+
+        const double mu_rho = mArtificialMassDiffusivityPrandtl * mu_h / inf_norm.Density;
+        const double kappa  = mArtificialConductivityPrandtl * mu_h / (heat_capacity_ratio - 1.0);
+
+        DistributeVariablesToNodes(r_element, mu_h, mu_rho, kappa);
+    });
+
+    KRATOS_CATCH("")
+}
+
+
+void ShockCapturingEntropyViscosityProcess::DistributeVariablesToNodes(
+    Element& rElement,
+    const double ArtificialBulkViscosity,
+    const double ArtificialMassDiffusivity,
+    const double ArtificialConductivity) const
+{
+    auto& r_geometry = rElement.GetGeometry();
+    const double element_volume = r_geometry.LocalSpaceDimension() == 3 ? r_geometry.Volume() : r_geometry.Area();
+
+    for(unsigned int i=0; i<r_geometry.size(); ++i)
+    {
+        auto& r_node = r_geometry[i];
+        const double weight = element_volume / r_node.GetValue(NODAL_AREA);
+        r_node.SetLock();
+        r_node.GetValue(ARTIFICIAL_BULK_VISCOSITY) += weight * ArtificialBulkViscosity;
+        r_node.GetValue(ARTIFICIAL_MASS_DIFFUSIVITY) += weight * ArtificialMassDiffusivity;
+        r_node.GetValue(ARTIFICIAL_CONDUCTIVITY) += weight * ArtificialConductivity;
+        r_node.UnSetLock();
+    }
+}
+
+
+double ShockCapturingEntropyViscosityProcess::ComputeEntropy(
+    const double Density,
+    const double Pressure,
+    const double Gamma)
+{
+    return Density / (Gamma - 1.0) * std::log(Pressure / std::pow(Density, Gamma));
+}
+
+
+double ShockCapturingEntropyViscosityProcess::ComputeHSquared(const Element& rElement)
+{
+    double h_squared = std::numeric_limits<double>::max();
+    const auto& r_geometry = rElement.GetGeometry();
+
+    // H is the shortest edge of the element
+    for(unsigned int i=0; i<r_geometry.size(); ++i)
+    {
+        const unsigned j = (i + 1) % r_geometry.size();
+        const auto edge = r_geometry[j] - r_geometry[i];
+        h_squared = std::min(h_squared, inner_prod(edge, edge));
+    }
```

Comment:
I think this is equivalent to the `MinimumElementSize` in the `ElementSizeCalculator`.

Reply:
I'm unsure. The article I based this on specifically mentioned the shortest edge of the element. I believe MinimumElementSize computes the shortest diameter, which will be smaller or equal to the shortest side.

Reply:
I tried it out with the green triangle an I get:
- `MinimumElementSize = 0.6`  (= the black line)
- `ComputeHSquared = 1.0` (=the bottom edge)

![image](https://user-images.githubusercontent.com/47142856/143269190-173dac7c-6f0e-453a-922d-e46a264bd9ec.png)

I don't believe this would have too big an impact on the result but they are diferent.